### PR TITLE
Enc service endpoint

### PR DESCRIPTION
**Pull request by matas-cc**
### Description
- Previously removed Encryption service reintroduced with two new endpoints - Encrypt, Decrypt.
- Added new service, registering it in gRPC.
- Added ECCS methods and tests for it.
- Authz handler interface and functions moved outside storage into a separate service. Authz implementation imports renamed from authz to authzimpl.

### Parent Issue
https://github.com/cyber-crypt-com/encryptonize-core/issues/195

### Comments for the Reviewers

- enc.proto EncryptResponse and DecryptRequest contain the same properties. This is intended, but maybe they should be extracted into a new type?
- ECCS Json and Base64 usage. Currently encoding ciphertext and aad to base64 and formatting to Json as part of the app. Maybe it should be part of tests?

### Type(s) of Change (Split the PR if you check many)
- [x] Added user feature
- [ ] Added technical feature
- [x] Added tests
- [ ] Refactor
- [ ] Bugfix
- [ ] Updated documentation
- [ ] Infrastructure changes (Terraform, GCP, K8s, other)
- [ ] CI/CD workflow changes

### Testing Checklist
- [x] I have added/updated unit tests for functional changes.
- [ ] I have added/updated integration tests for changes that affect dependent systems.
- [x] I have added/updated end-to-end tests for feature changes.

### General Checklist
- [x] I have pulled in the latest changes from master.
- [x] I have run `make lint`.
- [x] I have successfully run `make docker-up; make tests`.
- [ ] I have updated relevant workflows and deployment tools.
- [ ] I have updated/added relevant documentation (readme, doc comments, etc).
- [ ] I have added the license to new source code files.
- [x] I have spent some time looking over the full diff before creating this PR.

### Technical Debt
Configuration of service enabling and documentation to be addressed in 
https://github.com/cyber-crypt-com/encryptonize-core/issues/194
https://github.com/cyber-crypt-com/encryptonize-core/issues/196

### CODE REVIEW

File: encryption-service/services/enc/enc_handlers.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+	accessObject, err := authz.AuthorizeWrapper(ctx, enc.Authorizer, objectIDString)
+	if err != nil {
+		// AuthorizeWrapper logs and generates user facing error, just pass it on here
+		return nil, err
+	}
+
+	plaintext, err := enc.DataCryptor.Decrypt(accessObject.GetWOEK(), request.Ciphertext, request.AssociatedData)
+	if err != nil {
+		log.Error(ctx, err, "Retrieve: Failed to decrypt object")
+		return nil, status.Errorf(codes.Internal, "error encountered while retrieving object")
```
Comment (philip-cc) [inline encryption-service/services/enc/enc_handlers.go:81]:
```suggestion
		log.Error(ctx, err, "Decrypt: Failed to decrypt object")
		return nil, status.Errorf(codes.Internal, "error encountered while decrypting object")
```

File: encryption-service/services/enc/enc_handlers_test.go [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+import (
+	"bytes"
+	context "context"
+	"encryption-service/contextkeys"
+	"encryption-service/impl/authstorage"
+	authzimpl "encryption-service/impl/authz"
+	"encryption-service/impl/crypt"
+	"testing"
+
+	"github.com/gofrs/uuid"
```
Comment (philip-cc) [inline encryption-service/services/enc/enc_handlers_test.go:12]: Please group these correctly.

File: encryption-service/services/enc/enc_handlers_test.go [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+	if err != nil {
+		t.Fatalf("Encrypting object failed: %v", err)
+	}
+
+	decryptResponse, err := enc.Decrypt(
+		ctx,
+		&DecryptRequest{
+			ObjectId:       encryptResponse.ObjectId,
+			Ciphertext:     encryptResponse.Ciphertext,
+			AssociatedData: associatedData,
```
Comment (philip-cc) [inline encryption-service/services/enc/enc_handlers_test.go:60]:
```suggestion
			AssociatedData: encryptResponse.AssociatedData,
```

File: encryption-service/services/enc/enc_handlers_test.go [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+	)
+
+	if err != nil {
+		t.Fatalf("Decrypting object failed: %v", err)
+	}
+
+	comp := bytes.Compare(decryptResponse.Plaintext, plaintext)
+	if comp != 0 {
+		t.Fatalf("Decrypted plaintext does not equal original plaintext!")
+	}
```
Comment (philip-cc) [inline encryption-service/services/enc/enc_handlers_test.go:71]: We could also have a check that the associated data returned in both responses is correct.

File: encryption-service/services/enc/enc_handlers_test.go [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+	if err != nil {
+		t.Fatalf("Could not create user ID: %v", err)
+	}
+
+	ctx := context.WithValue(context.Background(), contextkeys.UserIDCtxKey, userID)
+
+	_, err = enc.Decrypt(ctx, fakeRequest)
+	if err == nil {
+		t.Fatalf("Decrypt should have errored")
+	}
```
Comment (philip-cc) [inline encryption-service/services/enc/enc_handlers_test.go:100]: I'd suggest the following additonal tests:
* Decrypt correct ciphertext with wrong AD.
* Decrypt with the OID of another object.

File: applications/ECCS/app/encrypt.go [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+	if filename != "" {
+		plaintext = openFile(filename)
+	}
+	if stdin {
+		data, err := io.ReadAll(os.Stdin)
+		if err != nil {
+			log.Fatalf("%v: %v", utils.Fail("Encrypt failed"), err)
+		}
+		plaintext = data
+	}
```
Comment (philip-cc) [inline applications/ECCS/app/encrypt.go:51]: This seems to be duplicated quite a lot. Perhaps a utility function is in order.

File: encryption-service/services/enc/enc_handlers_test.go (hunk 2 above)
Comment (matas-cc) [reply encryption-service/services/enc/enc_handlers_test.go]: Fixed

File: applications/ECCS/app/encrypt.go (hunk 6 above)
Comment (matas-cc) [reply applications/ECCS/app/encrypt.go]: Added a `readInput` function that's used in Store, Encrypt, Decrypt

File: encryption-service/impl/crypt/cryptor_test.go [hunk 7]
```diff
@@ -118,3 +118,44 @@ func TestAESCrypterEncryption(t *testing.T) {
 		t.Fatalf("ciphertext doesn't match:\n%x\n!=\n%x", expectedPlaintext, plaintext)
 	}
 }
+
+func TestAESCrypterEncryptWithKey(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (matas-cc) [reply encryption-service/impl/crypt/cryptor_test.go:122]: Added a test `TestAESCrypterInvalidLength`

File: encryption-service/services/enc/enc_handlers_test.go (hunk 5 above)
Comment (matas-cc) [reply encryption-service/services/enc/enc_handlers_test.go:104]: Created the tests for wrong AD and ODI

File: encryption-service/services/enc/enc_handlers_test.go (hunk 4 above)
Comment (matas-cc) [reply encryption-service/services/enc/enc_handlers_test.go:82]: Good point, added

**Review by philip-cc — APPROVED** [review]
Ship it!

**Pull request merged**